keps/sig-apps/3850-backoff-limits-per-index-for-indexed-jobs/README.md

```diff
@@ -0,0 +1,1165 @@
+kind: Job
+spec:
+  parallelism: 10
+  completions: 10
+  completionMode: Indexed
+  runPolicy:
+    backoffLimitPerIndex: true
+    maxFailedIndexes: 5
+  template:
+    spec:
+      restartPolicy: Never
+      containers:
+      - name: job-container
+        image: job-image
+        command: ["./tests-runner"]
+```
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+
+#### The Job object too big
+
+With the new field `.status.failedIndexes` the Job object can be significantly
+larger as every failed index is recorded in the field. For example, using the
+same format as for the `.status.completedIndexes` and allowing for completions
+equal to `10^6` the size of the field, in the worst-case scenario is 3.4MB, which
+could put significant overhead on fetching the Job object data from etcd.
+
+In order to mitigate this risk we constrain the maximal value of the
+`.spec.completions` field to `10^5`, when `.spec.runPolicy.backoffLimitPerIndex=true`.
+With this limit the size of the field in the worst-case scenario (of
+non-consecutive integers from `0` to `.spec.completions-1`) is limited by `295KB`.
+Even if the `.status.completedIndexes` reaches its limits, still the size of
+these two fields together will be limited by `690KB`, which is enough for etcd
+to handle.
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+We introduce a new Job API field, called `.spec.runPolicy.backoffLimitPerIndex`.
+When set to `true` it indicates that the `spec.backoffLimit` field is used to
+limit the number of retries, counted independently for all indexes.
+
+Additionally, we propose the `.spec.runPolicy.maxFailedIndexes` to control
+the maximal number of failed indexes. Once the number is exceeded the entire
+Job is marked Failed and its execution is terminated.
+
+We also propose to extend the PodFailurePolicy with a new action, called
+`FailIndex` to allow an index to fail fast before reaching the backoff limit
+per index.
+
+### Job API
+
+```golang
+
+// PodFailurePolicyAction specifies how a Pod failure is handled.
+// +enum
+type PodFailurePolicyAction string
+
+const (
+  // This is an action which might be taken on a pod failure - mark the
+  // Job's index as failed to avoid restarts within this index.
+  PodFailurePolicyActionFailIndex PodFailurePolicyAction = "FailIndex"
+  ...
+)
+...
+
+// JobRunPolicy describes custom policy for running a Job
+type JobRunPolicy struct {
```

Comment:
`PodFailurePolicy` sounds more related to pods, but `MaxFailedIndexes` is a concept for Job I think. But `PodFailurePolicy` is alpha now, maybe we can rename it to `FailurePolicy` and containing the other failure policies.

We can have a `completionPolicy` in parallel with `FailurePolicy`.

Reply:
@kerthcet IIRC, `podFailurePolicy` graduated to beta in v1.26. So renaming that's field might be hard.

https://kubernetes.io/docs/tasks/job/pod-failure-policy/#using-pod-failure-policy-to-ignore-pod-disruptions

Reply:
Yes, I think renaming of `podFailurePolicy` is hard, and I think what it currently does is also conceptually fitting its name. 

For the new fields (`backoffLimitPerIndex` and `maxFailedIndexes`) we have so far proposed:
1. `runPolicy` (as in KEP)
2. `completionPolicy` (I like it, very close to `completionMode: Indexed`)
3. `failurePolicy`

Out of the three I'm more or less equal between 1. and 2. 3. is slightly as it currently fits well, but I'm also considering that the field might be reused for Job success policies.

Reply:
> podFailurePolicy graduated to beta in v1.26

Then https://github.com/kubernetes/kubernetes/pull/117802

Reply:
Thanks for spotting and opening the PR fix!

Reply:
if we go with `runPolicy` or `completionPolicy` (they are similar enough that we can figure this out in the code review), then we should reuse it for kubernetes/kubernetes#117600

Reply:
sg

Reply:
@deads2k wdyt? 

The motivation to add `runPolicy` is that the new `backoffLimitPerIndex` field seems to introduce a new mode of operation to indexed jobs. The name is proposed generic so that we don't commit to indexed jobs.

Having said that, I think putting the new fields on the top level avoids the non-obvious naming decision for the field and puts the `backoffLimit` and `backoffLimitPerIndex` next to another, so +1 for moving them to the top level.

Reply:
ok with me

Reply:
Updated to make the new fields at the top level. Resolving the discussion.